msend.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.Dd Nov 11, 2024
.\" Please adjust this date whenever revising the manpage.
.Dt MSEND 8 SMM
.Os
.Sh NAME
.Nm msend
.Nd send UDP messages to a multicast group
.Sh SYNOPSIS
.Nm
.Op Fl 46hnvq
.Op Fl c Ar NUM
.Op Fl g Ar GROUP
.Op Fl join
.Op Fl i Ar ADDRESS
.Op Fl I Ar INTERFACE
.Op Fl p Ar PORT
.Op Fl P Ar PERIOD
.Op Fl t Ar TTL
.Op Fl text Ar 'text'
.Sh DESCRIPTION
Continuously send UDP packets to the multicast group specified by the
.Fl g
and
.Fl p
options.
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl 4
Select IPv4 test group, use with
.Fl I Ar IFACE
when not using
.Fl i Ar ADDRESS
or any
.Fl g Ar GROUP .
This is the default.
.It Fl 6
Select IPv6 test group when not using
.Fi i
or
.Fl g .
For an example, see below.
.It Fl c Ar NUM
Limit number of packets to send, default: unlimited.
.It Fl g Ar GROUP
Specify the IP multicast group address to which the packets are sent.
The default group for IPv4 is
.Nm 224.1.1.1
and for IPv6
.Nm ff2e::1 .
.It Fl p Ar PORT
Specify the UDP port number used by the multicast group.  The default
port number is
.Nm 4444 .
.It Fl join
Multicast sender will join join the multicast group.  By default, a
multicast sender does not (need to) join the group.
.It Fl t Ar TTL
Specify the TTL (1-255) value in the message.  You must increase this if
you want to route the traffic, otherwise the first router will drop the
packets!  The default value is 1.
.It Fl i Ar ADDRESS
Specify the IP address of the interface to be used to send the packets.
The default value is INADDR_ANY which implies that the default interface
selected by the system will be used.
.It Fl I Ar INTERFACE
The interface on which to send.  Can be specified as an alternative to
.Fl i .
.It Fl P Ar PERIOD
Specify the interval in milliseconds between two transmitted packets.
The default value is 1000 milliseconds.
.It Fl q
Quiet mode, do not log to stdout every time a message is successfully
sent.  Errors are stil logged.
.It Fl text Ar 'text'
Specify a message text which is sent as the payload of the packets and
is displayed by the
.Xr mreceive 8
command.  The default value is an empty string.
.It Fl n
Interpret the contents of the message text as a number instead of a
string of characters.  Use
.Nm mreceive
.Fl n
on the other end to interpret the message text correctly.
.It Fl v
Print version information.
.It Fl h
Print the command usage.
.El
.Sh EXAMPLE
.Bd -literal -offset left
$ msend -6 -I eth0
Now sending to multicast group: [ff2e::1]:4444
Sending msg 1, TTL 1, to [ff2e::1]:4444:
Sending msg 2, TTL 1, to [ff2e::1]:4444:
Sending msg 3, TTL 1, to [ff2e::1]:4444:
Sending msg 4, TTL 1, to [ff2e::1]:4444:
Sending msg 5, TTL 1, to [ff2e::1]:4444:
...
.Ed
.Sh SEE ALSO
.Xr mreceive 8
.Sh AUTHORS
.An -nosplit
.Nm mtools ,
originally called mSendReceive, was written by
.An Jianping Wang ,
.An Yvan Pointurier ,
and
.An Jörg Liebeherr
while at University of Virginia's Multimedia Networks Group.  It is
currently maintained by
.An Joachim Wiberg
at
.Lk https://github.com/troglobit/mtools "GitHub" .


ChangeLog.md
Change Log
==========

All relevant changes are documented in this file.


[v3.1][] - 2024-11-11
---------------------

### Changes
- Add optional `-s ADDR` argument for SSM to `mreceive`, issue #3
- Add support for quiet (`-q`) mode operation
- Add credit to original authors in README and man pages
- Refactor the option parsers to use `getopt(3)`
- Major refactoring and coding style cleanup, splitting the code base up
  in more files, renaming internal APIs, all to be able to use RFC3678
  style APIs for source filtering, issue #3

### Fixes
- Boundary checking, max number of interfaces, internal buffer lengths
- Fixes for building on non-Linux systems, tested on FreeBSD 11.0


[v3.0][] - 2024-02-24
---------------------

This release adds IPv6 support, contributed by Vladimir Oltean.

### Changes

- Add `-I INTERFACE` option to bind to a specific interface for send/receive
- Add `-c NUM` option to only send a limited number of packets
- Add `-4` and `-6` options to select default group to send/receive with `I`
- Sort command line options alphabetically
- Update documentation

### Fixes

- Fix `-v` to print correct version instead of hard-coded old version
- Bind to port after binding to device to receive only test packets


[v2.3][] - 2015-07-30
---------------------

First release of the mtools multicast testing suite (`msend` and `mreceive`)
after importing the sources from the [Multimedia Networks Group, University of
Virginia](http://www.cs.virginia.edu/~mngroup/software/).

> **Note:** First effort to adapt `ttcp.c` as well, but not tested yet,
> so not part of the build/install system.

### Changes

- Code cleanup and modernization to be able to build with a modern
  (2015) GCC and Clang
- Reindent code to use KNF
- Simplifiy and clean up online, -h, command usage help
- Removed old support for Windows.  May add Cygwin support in the future
- Cleanup of documentation, add common README and create real man pages:
  `msend.8` and `mreceive.8`
- Create real public domain LICENSE file
- Tested on Ubuntu Linux 14.04 (Amd64) and FreeBSD (Amd64)


[v3.1]: https://github.com/troglobit/mtools/compare/v3.0...v3.1
[v3.0]: https://github.com/troglobit/mtools/compare/v2.3...v3.0
[v2.3]: https://github.com/troglobit/mtools/compare/v2.2...v2.3
[v2.2]: https://github.com/troglobit/mtools/compare/v2.1...v2.2
[v2.1]: https://github.com/troglobit/mtools/compare/v2.0...v2.1


msend.c
/*
 * msend.c  -- Sends UDP packets to a multicast group
 * 
 * (c)  Jianping Wang, Yvan Pointurier, Jorg Liebeherr, 2002
 *      Multimedia Networks Group, University of Virginia
 *
 * SOURCE CODE RELEASED TO THE PUBLIC DOMAIN
 * 
 * version 2.0 - 5/20/2002 
 * version 2.1 - 12/4/2002  
 * 	By default, msend does not join multicast group. If  -join option is 
 * 	given, msend joins the multicast group. 
 * version 2.2 - 05/17/2003  
 *      Most commandline parameters are assigned default values. The 
 *      usage information is changed according to README_msend.txt
 * 
 * 
 * Based on this public domain program:
 * u_mctest.c            (c) Bob Quinn           2/4/97
 * 
 */

#include "common.h"


static int usage(int rc)
{
	printf("\
Usage:  msend [-46hnv] [-c NUM] [-g GROUP] [-p PORT] [-join] [-i ADDRESS]\n\
	      [-I INTERFACE] [-P PERIOD] [-t TTL] [-text \"text\"]\n\
\n\
  -4 | -6      Select IPv4 or IPv6, use with -I, when -i is not used\n\
  -c NUM       Number of packets to send. Default: send indefinitely\n\
  -g GROUP     IP multicast group address to send to.\n\
               Default for IPv4: 224.1.1.1, IPv6: ff2e::1\n\
  -h           This help text.\n\
  -i ADDRESS   IP address of the interface to use to send the packets.\n \
               The default is to use the system default interface.\n\
  -I INTERFACE The interface on which to send. Can be specified as an\n\
               alternative to -i.\n\
  -join        Multicast sender will join the multicast group.\n\
               By default a sender never joins the group.\n\
  -n           Encode -text argument as a number instead of a string.\n\
  -p PORT      UDP port number used in the multicast packets.  Default: 4444\n\
  -P PERIOD    Interval in milliseconds between packets.  Default 1000 msec\n\
  -q           Quiet, don't print 'Sedning msg ...' for every packet\n\
  -t TTL       The TTL value (1-255) used in the packets.  You must set\n\
               this higher if you want to route the traffic, otherwise\n\
               the first router will drop the packets!  Default: 1\n\
  -text \"text\" Specify a string to use as payload in the packets, also\n\
               displayed by the mreceive command.  Default: empty\n\
  -v           Print version information.\n\n");

	return rc;
}

static int do_send(int sd, inet_addr_t *to, char *msg, size_t len, int isnum)
{
	static int counter = 1;
	int ret;

	if (isnum)
		snprintf(msg, len, "%d", counter);

	ret = sendto(sd, msg, len, 0, (struct sockaddr *)to, inet_addrlen(to));
	if (ret < 0) {
		perror("sendto");
		exit(1);
	}

	return counter++;
}

int main(int argc, char *argv[])
{
	static struct option opts[] = {
		{ "join",       no_argument,       NULL, 'j' },
		{ "text",       required_argument, NULL, 'T' },
		{ NULL,         0,                 NULL, 0   }
	};
	inet_addr_t ifaddr, group;
	char msg[BUFSIZE] = { 0 };
	int ret, c, sd;

	while ((c = getopt_long_only(argc, argv, "46c:g:hi:I:jnp:P:qt:T:v", opts, NULL)) != EOF) {
		switch (c) {
		case '4':
			opt_family = AF_INET; /* for completeness */
			break;
		case '6':
			opt_family = AF_INET6;
			break;
		case 'c':
			opt_count = atoi(optarg);
			break;
		case 'g':
			group_addr = optarg;
			break;
		case 'h':
			return usage(0);
		case 'i':
			if (opt_ifaddr) {
				fprintf(stderr, "Single source address allowed\n");
				exit(1);
			}
			opt_ifaddr = strdup(optarg);
			if (!opt_ifaddr) {
				perror("strdup");
				exit(1);
			}
			break;
		case 'I':
			if (opt_ifname) {
				fprintf(stderr, "Single interface expected\n");
				exit(1);
			}
			opt_ifname = optarg;
			break;
		case 'j':
			opt_join++;
			break;
		case 'n':
			opt_isnum = 1;
			break;
		case 'p':
			group_port = atoi(optarg);
			break;
		case 'P':
			opt_period = atoi(optarg);
			break;
		case 'q':
			opt_verbose = 0;
			break;
		case 't':
			opt_ttl = atoi(optarg);
			break;
		case 'T':
			strlcpy(msg, optarg, sizeof(msg));
			break;
		case 'v':
			printf("msend version %s\n", VERSION);
			return 0;
		default:
			fprintf(stderr, "wrong parameters!\n\n");
			return usage(1);
		}
	}

	if (!opt_ifaddr) {
		if (opt_family == AF_INET)
			opt_ifaddr = "0.0.0.0";
		else
			opt_ifaddr = "::";
	}

	ret = inet_parse(&ifaddr, opt_ifaddr, group_port);
	if (ret) {
		fprintf(stderr, "IP address %s not in known format\n", opt_ifaddr);
		exit(1);
	}
	opt_family = ifaddr.ss_family;

	if (group_addr == NULL) {
		if (opt_family == AF_INET)
			group_addr = TEST_ADDR_IPV4;
		else if (opt_family == AF_INET6)
			group_addr = TEST_ADDR_IPV6;
		else
			exit(1);
	}

	ret = inet_parse(&group, group_addr, group_port);
	if (ret) {
		fprintf(stderr, "Group address %s not in known format\n", group_addr);
		exit(1);
	}

	if (opt_join && group.ss_family == AF_INET6 && !opt_ifname) {
		fprintf(stderr, "-I is mandatory when joining IPv6 group\n");
		exit(1);
	}

	/* get a datagram socket */
	sd = sock_create(&ifaddr, opt_ifname);
	if (sd < 0)
		exit(1);

	/* join the multicast group we are sending to (usually not necessary!) */
	if (opt_join) {
		ret = sock_mc_join(sd, NULL, &group, opt_ifname, 0, NULL);
		if (ret)
			exit(1);
	}

	/* set TTL to traverse up to multiple routers */
	ret = sock_mc_ttl(sd, opt_ttl);
	if (ret)
		exit(1);

	/* enable loopback */
	ret = sock_mc_loop(sd, 1);
	if (ret)
		exit(1);

	logit("Now sending to multicast group: [%s]:%d\n", group_addr, group_port);

	opt_period *= 1000;	/* convert to microsecond */
	if (opt_period > 0) {
		struct itimerval it;
		sigset_t set;

		/* block SIGALRM */
		sigemptyset(&set);
		sigaddset(&set, SIGALRM);
		sigaddset(&set, SIGINT);
		sigprocmask(SIG_BLOCK, &set, NULL);

		/*
		 * set up interval timer, sends SIGALRM every opt_period msec
		 */
		it.it_value.tv_sec     = 0;	/* wait a bit for system to "stabilize"  */
		it.it_value.tv_usec    = 1;	/* tv_sec or tv_usec cannot be both zero */
		it.it_interval.tv_sec  = (time_t)(opt_period / 1000000);
		it.it_interval.tv_usec =   (long)(opt_period % 1000000);
		setitimer(ITIMER_REAL, &it, NULL);

		for (;;) {
			int signo = sigwaitinfo(&set, NULL);

			if (signo == SIGALRM) {
				ret = do_send(sd, &group, msg, sizeof(msg), opt_isnum);
				logit("Sending msg %d, TTL %d, to [%s]:%d: %s\n", ret, opt_ttl, group_addr, group_port, msg);
			} else {
				logit("Git signal %d, exiting!\n", signo);
				break;
			}

			if (ret == opt_count)
				break;
		}

		return 0;
	} else {
		for (int i = 0; i < opt_count; i++) {
			do_send(sd, &group, msg, sizeof(msg), opt_isnum);
			logit("Send out msg %d to [%s]:%d: %s\n", i, group_addr, group_port, msg);
		}
	}

	return 0;
}

/**
 * Local Variables:
 *  c-file-style: "linux"
 *  indent-tabs-mode: t
 * End:
 */
